_drafts/code-properties-and-language-features.md

```diff
@@ -0,0 +1,50 @@
+
+It's difficult for me to find answers to these questions. As Yehuda Katz [tweeted][yehuda-1] [recently][yehuda-2], there's a tendency to describe programming advancements as radically new paradigms, and past approaches as terrible. There's no grappling with the arguments older approaches made.
+
+This leaves me with a conundrum: I bought into arguments about encapsulation increasing maintainability. Were those arguments true, or were they just something we made up to tell ourselves so we could build software in a way we liked?
+
+Even though these things aren't discussed much publicly, asking questions of people who know Elixir and Redux helped me with this.
+
+# Questioning Elixir and Redux
+
+First, I asked about the way data is separate from functionality in Elixir, and whether that caused problems. But I learned that when you define a struct inside a module, you also usually define functions that operate on that struct inside the same module. Because the name of the module and the name of the struct are the same, they go together naturally. So data and functionality aren't entirely separated.
+
+Then I asked about encapsulation: the data in a struct is available for everyone else in the app to access directly. But I learned that often you might have a convention of only accessing data in the struct via a function, and that that is a way to isolate knowledge of the data structure.
+
+Then I asked about [the `Plug.Conn` struct][conn] in the Phoenix framework, which is passed through much of the app. I asked about whether this made the entire application fragile, because all of it required intimate knowledge of the Connection. I learned that there was a [`private`][conn-private] field on the struct, and although you aren't prevented from accessing it directly, the convention is that you shouldn't because the implementation in it can change without warning. This reminded me of some Rubyists' point that, because you can get around the private keyword, it's effectively just a convention as well.
+
+After all this, I still had a concern with [Elixir's pattern matching][pattern-matching]. It's a lot like a switch statement, and [switch statements are a code smell][switch] in OO design. It makes it difficult to change your code because switch statements proliferate: every time you have to add a new case, all the switch statements need to be updated. Instead it's best to use polymorphism to handle variations, so that you can always add a new type of object and implement the new case's behavior on it. Each place you previously had a switch statement, you now send a message instead. But Elixir has a potential answer to this as well: [protocols][protocols] allow you to define a function signature, and then you can implement it for specific data types.
```

Comment:
@iamvery Great observation; I just checked the Elixir pattern-matching docs and they don't list switching on struct types at all. I think the "switch statement" code smell name was misleading! Maybe that code smell could better be named "switching on types".

I'll update the article to reference this misunderstanding.